gowin/globals.cc

```diff
@@ -0,0 +1,335 @@
+
+            // >>> GBOx <- GTx0
+            dstWire = ctx->getPipSrcWire(gb_pip_id);
+            WireInfo dstWireInfo = ctx->wire_info(dstWire);
+            int branch_tap_idx = clock > 3 ? 1 : 0;
+            snprintf(buf, sizeof(buf), "R%dC%d_GT%d0_GBO%d", dstWireInfo.y + 1, dstWireInfo.x + 1, branch_tap_idx,
+                     branch_tap_idx);
+            PipId gt_pip_id = ctx->id(buf);
+            if (ctx->verbose) {
+                log_info("     GT Pip:%s\n", buf);
+            }
+            // sanity
+            NPNR_ASSERT(find(ctx->getPipsUphill(dstWire).begin(), ctx->getPipsUphill(dstWire).end(), gt_pip_id) !=
+                        ctx->getPipsUphill(dstWire).end());
+            // if already routed
+            if (used_pips.count(gt_pip_id)) {
+                if (ctx->verbose) {
+                    log_info("     ^routed already^\n");
+                }
+                continue;
+            }
+            used_pips.insert(gt_pip_id);
+
+            // >>> GTx0 <- SPINExx
+            // XXX no optimization here, we need to store
+            // the SPINE <-> clock# correspondence in the database. In the
+            // meantime, we define in run-time in a completely suboptimal way.
+            std::vector<std::string> clock_spine;
+            dstWire = ctx->getPipSrcWire(gt_pip_id);
+            for (auto const &uphill_pip : ctx->getPipsUphill(dstWire)) {
+                std::string name = ctx->wire_info(ctx->getPipSrcWire(uphill_pip)).type.str(ctx);
+                if (name.rfind("SPINE", 0) == 0) {
+                    clock_spine.push_back(name);
+                }
+            }
+            sort(clock_spine.begin(), clock_spine.end(), [](const std::string &a, const std::string &b) -> bool {
+                return (a.size() < b.size()) || (a.size() == b.size() && a < b);
+            });
+            dstWireInfo = ctx->wire_info(dstWire);
+            snprintf(buf, sizeof(buf), "R%dC%d_%s_GT%d0", dstWireInfo.y + 1, dstWireInfo.x + 1,
+                     clock_spine[clock - branch_tap_idx * 4].c_str(), branch_tap_idx);
+            PipId spine_pip_id = ctx->id(buf);
+            if (ctx->verbose) {
+                log_info("      Spine Pip:%s\n", buf);
+            }
+            // sanity
+            NPNR_ASSERT(find(ctx->getPipsUphill(dstWire).begin(), ctx->getPipsUphill(dstWire).end(), spine_pip_id) !=
+                        ctx->getPipsUphill(dstWire).end());
+            // if already routed
+            if (used_pips.count(spine_pip_id)) {
+                if (ctx->verbose) {
+                    log_info("      ^routed already^\n");
+                }
+                continue;
+            }
+            used_pips.insert(spine_pip_id);
+
+            // >>> SPINExx <- IO
+            dstWire = ctx->getPipSrcWire(spine_pip_id);
+            dstWireInfo = ctx->wire_info(dstWire);
+            PipId io_pip_id = PipId();
+            for (auto const &uphill_pip : ctx->getPipsUphill(dstWire)) {
```

Comment:
nitpick but because `PipId` is a small register-size type you don't need the `&` here

Reply:
What was I thinking of? Fixed.